### PR TITLE
Make STAR file data reads/writes thread-safe and eliminate spurious copies

### LINKED ISSUE #26: Datasets are not safe for multithreading!
There are some illegal things we do under the hood if one is to plug a dataset into a dataloader for multithreaded read/writes. It’s time we fix this!

### PR DESCRIPTION
**Pull request by michael-0brien**
This PR makes reading/writing data from the `RelionParticleParameterFile` thread-safe.

For writing, the issue was that we were not updating the dataframes in-place on calls to `dataset[index] = value` because of how `updates_optics_group` worked. 

For reading, we were briefly changing the internal state of the class during calls to `value = dataset[index]`. After this PR, it should be safe to wrap datasets into a dataloader.

### CODE REVIEW

**Comment by michael-0brien** [comment]
@DSilva27 it would be helpful to have feedback on API here (particularly on the new `max_optics_groups` parameter, e.g. what its default values should be). There was no perfect solution here, but the idea was to make it so different writes do independent in-place modifications to the internal dataframes.

**Comment by michael-0brien** [comment]
This refactor has ended up having some general usability adjustments and other optimizations. For example, `parameter_file.starfile_data` is chopped in favor of `parameter_file.particle_data` and `parameter_file.optics_data`, and setters for each are removed to guard against spurious updates (these can likely break the loader).

We can alternatively allow the user to make these reference updates and throw a warning, but thought I would keep things simple to start.

**Comment by michael-0brien** [comment]
I think it's working well now! This should give serious improvements on STAR file reading/writing. However, I bet there's still bugs in edge cases; I'll mark it ready for review but need to write tests.

**Comment by DSilva27** [comment]
> This refactor has ended up having some general usability adjustments and other optimizations. For example, `parameter_file.starfile_data` is chopped in favor of `parameter_file.particle_data` and `parameter_file.optics_data`, and setters for each are removed to guard against spurious updates (these can likely break the loader).
> 
> We can alternatively allow the user to make these reference updates and throw a warning, but thought I would keep things simple to start.

I agree with this. In practice changes would be better done by creating a new parameter file in writing mode. In the future we could have a "rw" mode that allows someone to use the setters, but I don't think it's a priority.

File: cryospax/_dataset/relion.py [hunk 1]
```diff
@@ -1154,20 +1303,62 @@ def only_images(self) -> bool:
...
             starfile_data = read_starfile(path_to_starfile)
             _validate_starfile_data(starfile_data)
+            # Handle particle entries
             if len(selection_filter) > 0:
                 starfile_data = _select_particles(starfile_data, selection_filter)
+            # Handle optics group entries
+            optics_data = starfile_data["optics"]
+            num_optics_groups, max_optics_group_index = (
+                len(optics_data),
+                optics_data["rlnOpticsGroup"].max().item(),
```
Comment (DSilva27) [inline cryospax/_dataset/relion.py]: I'm getting an error here when there is only one optics group in the starfile. If that's the case, then max() returns an integer, and item() fails.

Comment (DSilva27) [reply cryospax/_dataset/relion.py]: Note: this happens when I try to load the starfile in tests/ why are the tests not failing for the PR check?

**Review by DSilva27 — CHANGES_REQUESTED** [review]

File: cryospax/_dataset/relion.py [hunk 2]
```diff
@@ -2000,19 +2153,12 @@ def _make_image_filename(
...
-    except Exception as err:
-        raise OSError(
-            f"Could not get the file number from file {filename} "
-            "Files must be enumerated with the trailing part of the "
-            "filename as the file number, like so: '/path/to/file-0000.txt'. "
-            f"When extracting the file number and converting it to an integer, "
-            f"found error:\n\t{err}"
-        )
-    return file_number
+def _get_maximum_file_index(rln_image_name) -> int:
```
Comment (DSilva27) [inline cryospax/_dataset/relion.py:2156]: This fails when .mrcs are called `/path/to/data/prefix_NUMBER_suffix.mrc`. This happens, for example, when the output is a downsample job from cryosparc.

Comment (DSilva27) [reply cryospax/_dataset/relion.py:2156]: Fixed it with Claude, and it works for my usecases: `filenames.str.extract(r"(\d+)(?:_[^_.]+)?\.[^.]+$", expand=False)`

File: cryospax/_dataset/relion.py [hunk 3]
```diff
@@ -556,9 +638,14 @@ def save(
                 f"`{type(self).__name__}.overwrite = True`."
             )
         else:
+            # Make directory if it does not exist
             if not path_to_starfile.parent.exists():
                 path_to_starfile.parent.mkdir(parents=True)
-            write_starfile(self.starfile_data, path_to_starfile, **kwargs)
+            starfile_data = _StarfileData(
```
Comment (DSilva27) [inline cryospax/_dataset/relion.py:644]: Let's switch the order of these two so that the optics parameters are written first

File: pyproject.toml
Comment (michael-0brien) [inline pyproject.toml:1]: should be >=

File: cryospax/_dataset/relion.py (hunk 2 above)
Comment (michael-0brien) [reply cryospax/_dataset/relion.py:2156]: gotcha. what is the new syntax doing? it’s possible this could break something for real data, where collection software could possibly add multiple numeric markers

**Comment by michael-0brien** [comment]
Should be good for merge!

**Review by DSilva27 — APPROVED** [review]
